clients/src/main/java/org/apache/kafka/common/utils/Exit.java

```diff
@@ -30,16 +33,35 @@
         void addShutdownHook(String name, Runnable runnable);
     }
 
+    private static final Set<String> JUNIT_CLASSS_NAMES = new HashSet<>();
+
+    static {
+        JUNIT_CLASSS_NAMES.add("org.junit.platform.commons.util.ReflectionUtils");
+        JUNIT_CLASSS_NAMES.add("org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor");
+    }
+
+    static void throwIfInJunitTest() {
+        for (StackTraceElement[] elements : Thread.getAllStackTraces().values()) {
+            for (StackTraceElement element : elements) {
+                if (JUNIT_CLASSS_NAMES.contains(element.getClassName())) {
+                    throw new RuntimeException("Attempted to terminate the VM in a junit test.");
+                }
+            }
+        }
+    }
```

Comment:
Is the intention to allow termination to happen from other threads even if still running a test?
i.e. do we want this test to pass as well?

```scala
  @Test
  def exitWithoutSettingUpHooks(): Unit = {
    val t = new Thread {
      override def run {
        try {
          Exit.exit(1, None)
        } catch {
          case e: RuntimeException => assertEquals("Attempted to terminate the VM in a junit test.", e.getMessage)
        }
      }
    }
    t.start()
    t.join()
  }
```

It's also maybe slightly nicer to not have to check the stack frames

```suggestion
    private static final boolean IN_JUNIT_TEST = Stream.of(
            "org.junit.platform.commons.util.ReflectionUtils",
            "org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor"
    ).anyMatch(clazz -> {
        try {
            Class.forName(clazz);
            return true;
        } catch (ClassNotFoundException ignored) {
            return false;
        }
    });

    static void throwIfInJunitTest() {
        if (IN_JUNIT_TEST) {
            throw new RuntimeException("Attempted to terminate the VM in a junit test.");
        }
    }
```

Reply:
> Is the intention to allow termination to happen from other threads even if still running a test?

Why would we want this?

> (proposed changes)

It seems like somebody could end up with the junit jars on their CLASSPATH, but not be running a test. So I think looking at the stack trace is safer.